### PR TITLE
program: Updates for better 1 SOL minimum delegation support

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

The version of solana-program-test used by the repo doesn't properly add the BPF version of the stake program in tests -- if you take the current main branch, use the BPF stake program currently on mainnet, and run the tests, you'll see the minimum delegation reported is just 1 lamport.

On top of that, with the 1 SOL minimum delegation, the minimum limits used during `withdraw_stake` to figure out if one of the edge cases apply is too strict. It currently requires that the stake account withdrawn from has at least `rent_exemption + minimum_delegation + 1`, assuming that we can split 1 lamport from a stake account.

Since we need to split at least 1 SOL, we can have a situation where a stake account has `rent_exemption + 1_999_999_999` lamports. The stake pool program forces you to withdraw from that account, but it isn't possible.

#### Summary of changes

The first two commits just bump everything to the newest dependencies, update the stake program fixture, and use it correctly in program-test. Note that we have to use v4.1 because of the incompatible wincode bump in the SDK.

After that, modify `withdraw_stake` to use a looser bound of `rent_exemption + minimum_delegation * 2 + lamports_per_pool_token` as the limit where a user can withdraw from transient accounts / remove validators.

Only one existing test is tripped by the new behavior, so fix that, and add a new one.

### CODE REVIEW

File: program/src/processor.rs [hunk 1]
```diff
@@ -2952,11 +2958,9 @@ impl Processor {
                         ValidatorStakeInfo::memcmp_pubkey(x, &preferred_withdraw_validator)
                     })
                 {
-                    let available_lamports =
-                        u64::from(preferred_validator_info.active_stake_lamports)
-                            .saturating_sub(minimum_lamports_with_tolerance);
                     if preferred_withdraw_validator != vote_account_address
-                        && available_lamports > 0
+                        && u64::from(preferred_validator_info.active_stake_lamports)
+                            >= minimum_lamports_with_tolerance
```
Comment (joncinque) [inline program/src/processor.rs:2963]: I think `>=` is correct here, since if the stake account has exactly `minimum_lamports_with_tolerance`, it should be possible to withdraw a stake account

**Comment by joncinque** [comment]
#757 will fix the audit failure

File: program/src/processor.rs (hunk 1 above)
Comment (sintemal) [reply program/src/processor.rs:2963]: I agree, that >= is correct, but it is inconsistent with our previous check on `has_withdrawable_active_stake`, as there we check that it is >. 
So we can get into the situation, that `preferred_validator_info.active_stake_lamports == minimum_lamports_with_tolerance` and this guard will fire. But actually withdrawing this amount will not take the same path, as the `has_withdrawable_active_stake` is false for  `preferred_validator` , so it will go into validator removal, which is not actually necessary as far as I see.

File: program/tests/withdraw_edge_cases.rs [hunk 2]
```diff
@@ -1531,3 +1535,96 @@ async fn success_remove_preferred_validator_resets_preference() {
...
+    .await;
+    let error = stake_pool_accounts
+        .withdraw_stake(
+            &mut context.banks_client,
+            &context.payer,
+            &context.last_blockhash,
+            &user_stake_recipient.pubkey(),
+            &user_transfer_authority,
+            &deposit_info.pool_account.pubkey(),
+            &validator_stake.stake_account,
```
Comment (sintemal) [inline program/tests/withdraw_edge_cases.rs:1616]: this actually does not try to withdraw from a second non-preferred validator.

File: program/src/processor.rs (hunk 1 above)
Comment (joncinque) [reply program/src/processor.rs:2963]: Ah that's a great point, thanks for checking. I've made it `>=` in the latest commit [0095320](https://github.com/solana-program/stake-pool/pull/756/commits/00953206a50c9ade60f9d317445288573526ea79)

File: program/tests/withdraw_edge_cases.rs (hunk 2 above)
Comment (joncinque) [reply program/tests/withdraw_edge_cases.rs:1616]: That's correct, it's not really about the preferred behavior, it's more to make sure that we properly respect the required minimum.

I was mainly surprised that no tests broke after making the change 😅 I'll remove the preferred parts of this test, since they aren't really necessary

**Review by 2501babe — DISMISSED** [review]

**Review by 2501babe — APPROVED** [review]
